include/boost/geometry/extensions/multi/gis/io/wkb/detail/parser.hpp

```diff
@@ -0,0 +1,237 @@
+// Boost.Geometry (aka GGL, Generic Geometry Library)
+
+// Copyright (c) 2009-2012 Mateusz Loskot, London, UK.
+
+// Use, modification and distribution is subject to the Boost Software License,
+// Version 1.0. (See accompanying file LICENSE_1_0.txt or copy at
+// http://www.boost.org/LICENSE_1_0.txt)
+
+#ifndef BOOST_GEOMETRY_MULTI_IO_WKB_DETAIL_PARSER_HPP
+#define BOOST_GEOMETRY_MULTI_IO_WKB_DETAIL_PARSER_HPP
+
+#include <cassert>
+#include <cstddef>
+#include <iterator>
+#include <limits>
+
+#include <boost/geometry/extensions/gis/io/wkb/detail/endian.hpp>
+#include <boost/geometry/extensions/gis/io/wkb/detail/parser.hpp>
+#include <boost/geometry/extensions/gis/io/wkb/detail/ogc.hpp>
+
+
+#include <boost/geometry/multi/core/point_type.hpp>
+#include <boost/geometry/multi/core/ring_type.hpp>
+
+#include <boost/geometry/multi/core/interior_rings.hpp>
+
+namespace boost { namespace geometry
+{
+
+#ifndef DOXYGEN_NO_DETAIL
+namespace detail { namespace wkb
+{
+
+template <typename MultiPoint>
+struct multipoint_parser
+{
+    template <typename Iterator>
+    static bool parse(Iterator& it, Iterator end, MultiPoint& multipoint, byte_order_type::enum_t order)
+    {
+        if (!geometry_type_parser<MultiPoint>::parse(it, end, order))
+        {
+            return false;
+        }
+        
+        boost::uint32_t num_points(0);
+        if (!value_parser<boost::uint32_t>::parse(it, end, num_points, order))
+        {
+            return false;
+        }
+        
+        // Check that the number of double values in the stream is equal
+        // or greater than the number of expected values in the multipoint
+        
+        typedef typename std::iterator_traits<Iterator>::difference_type size_type;
+        
+        typedef typename point_type<MultiPoint>::type point_type;
+        
+        size_type const container_byte_size = dimension<point_type>::value * num_points * sizeof(double)
+                                            + num_points * sizeof(boost::uint8_t)
+                                            + num_points * sizeof(boost::uint32_t);
+        
+        size_type const stream_byte_size = std::distance(it,end);
+        
+        assert(stream_byte_size >= container_byte_size);
+        
+        point_type point_buffer;
+        std::back_insert_iterator<MultiPoint> output(std::back_inserter(multipoint));
```

Comment:
A MultiPoint may not have an interface of a STLContainer and in particular may not define `push_back()` member function, which means that `std::back_insert_iterator<>` may not work.

However in various places of the library, e.g. in set operations like `union_()` `std::back_inserter(out)` is passed into the internals of the algorithm. This implies that an Output must define `push_back()` member function which means that this algorithm (and others) may not work well with an arbitrary MultiGeometry. Since similar code can be found in other places in the library it can stay as it is. Nobody complained about it so I'm guessing the users are either using STL Containers, Boost.Geometry models or just implement `push_back()` member function.

We could think about a solution later, e.g. to implement our own `back_inserter`, but not as a part of this PR.